File: day1.js
const assert = require('assert')

// Part 1


const function_part_1 = (mass) => {
    return Math.floor(mass / 3) - 2
}


const test_function_part_1 = () => {
    const inputs = [12, 14, 1969, 100756]
    const outputs = [2, 2, 654, 33583]
    assert.deepEqual(outputs, inputs.map(function_part_1))
}

// Part 2


const function_part_2 = (mass, current_level) => {
    const fuel = function_part_1(mass)
    if (fuel <= 0) {
        return current_level
    }
    return function_part_2(fuel, current_level + fuel)
}


const test_function_part_2 = () => {
    const inputs = [14, 1969, 100756]
    const outputs = [2, 966, 50346]
    assert.deepEqual(outputs, inputs.map(function_part_2))
}


console.log("test 1 started")
test_function_part_1()
console.log("test 1 finished")
console.log("test 2 started")
test_function_part_2()
console.log("test 2 fished")

File: advent2019/day1.py
import math


def function_part_1(mass: int) -> int:
    return math.floor(mass / 3) - 2


def test_function_part_1():
    tests = [(12, 2), (14, 2), (1969, 654), (100756, 33583)]
    for mass, fuel in tests:
        assert function_part_1(mass) == fuel


def function_part_2(mass: int, current_level: int) -> int:
    fuel = function_part_1(mass)
    if fuel <= 0:
        return current_level
    return function_part_2(mass=fuel, current_level=current_level + fuel)


def test_function_part_2():
    tests = [(14, 2), (1969, 966), (100756, 50346)]
    for mass, fuel in tests:
        assert function_part_2(mass, 0) == fuel


if __name__ == '__main__':
    with open('data/input1.txt') as f:
        lines = [int(line) for line in f.readlines()]

    print('Part 1')
    print(sum(map(function_part_1, lines)))

    print('Part 2')
    print(sum([function_part_2(line, 0) for line in lines]))
